### ISSUE TITLE
new Bind master zone is empty

### ISSUE THREAD

**Issue by blablup** [opnsense/plugins#1467]
Hello,
If have a problem using bind as a authoritative name server for a domain. I created a domain and enabled it. The new domain gets inserted into named.conf correctly, but the create file at `/usr/local/etc/namedb/master/` has 0 bytes. 

The configd log part:

```
Aug 27 10:32:26 OPNsense configd.py: [1719a5da-31c9-454a-9108-cfc3a77f218f] generate template OPNsense/Bind 
Aug 27 10:32:26 OPNsense configd.py: generate template container OPNsense/Bind 
Aug 27 10:32:26 OPNsense configd.py:  OPNsense/Bind generated //usr/local/etc/namedb/master/bing.db 
Aug 27 10:32:26 OPNsense configd.py:  OPNsense/Bind generated //usr/local/etc/namedb/master/blacklist.db 
Aug 27 10:32:26 OPNsense configd.py:  OPNsense/Bind generated //usr/local/etc/namedb/master/test.dom.db 
Aug 27 10:32:26 OPNsense configd.py:  OPNsense/Bind generated //usr/local/etc/namedb/master/duckduckgo.db 
Aug 27 10:32:26 OPNsense configd.py:  OPNsense/Bind generated //usr/local/etc/namedb/master/google.db 
Aug 27 10:32:26 OPNsense configd.py:  OPNsense/Bind generated //etc/rc.conf.d/named 
Aug 27 10:32:26 OPNsense configd.py:  OPNsense/Bind generated //usr/local/etc/namedb/named.conf 
Aug 27 10:32:26 OPNsense configd.py:  OPNsense/Bind generated //usr/local/etc/namedb/rndc.conf 
Aug 27 10:32:26 OPNsense configd.py:  OPNsense/Bind generated //usr/local/etc/namedb/master/whitelist.db 
Aug 27 10:32:26 OPNsense configd.py:  OPNsense/Bind generated //usr/local/etc/namedb/whitelist.inc 
Aug 27 10:32:26 OPNsense configd.py:  OPNsense/Bind generated //usr/local/etc/namedb/master/youtube.db 
Aug 27 10:32:26 OPNsense configd.py: [7cf73bb7-0ef5-4f62-b483-cbc03d296c3e] request BIND status 
Aug 27 10:32:27 OPNsense configd.py: [e85fdcd7-4834-4bf5-ac41-393e6b45601b] starting BIND 
Aug 27 10:32:35 OPNsense configd.py: [e81d0af3-6912-45a8-8952-20ee96a9d0e0] request BIND status 

```

config.xml part:
```
   <bind>
      <acl version="1.0.0">
        <acls>
          <acl uuid="02ed2632-0d50-43e5-8f81-a500a9d403c2">
            <enabled>1</enabled>
            <name>myLocalnets</name>
            <networks>192.168.176.0/20</networks>
          </acl>
        </acls>
      </acl>
      <record version="1.0.0">
        <records>
        </records>
      </record>
      <dnsbl version="1.0.5">
        <enabled>1</enabled>
        <type>blp,emd,mw,pt,rw,za</type>
        <whitelists/>
        <forcesafegoogle>0</forcesafegoogle>
        <forcesafeduckduckgo>0</forcesafeduckduckgo>
        <forcesafeyoutube>0</forcesafeyoutube>
        <forcestrictbing>0</forcestrictbing>
      </dnsbl>
      <general version="1.0.5">
        <enabled>1</enabled>
        <enablerpz>1</enablerpz>
        <listenv4>127.0.0.1,192.168.178.225</listenv4>
        <listenv6>::1</listenv6>
        <port>53530</port>
        <forwarders>192.168.178.1</forwarders>
        <logsize>50</logsize>
        <maxcachesize>10</maxcachesize>
        <recursion>02ed2632-0d50-43e5-8f81-a500a9d403c2</recursion>
        <dnssecvalidation>no</dnssecvalidation>
        <hidehostname>1</hidehostname>
        <hideversion>1</hideversion>
        <enableratelimiting>1</enableratelimiting>
        <ratelimitcount>5</ratelimitcount>
        <ratelimitexcept>127.0.0.1,::1,192.168.176.0/20</ratelimitexcept>
      </general>
      <domain version="1.0.0">
        <domains>
          <domain uuid="e38802f4-1382-4921-9c62-6a5edb98221f">
            <enabled>1</enabled>
            <type>master</type>
            <masterip/>
            <domainname>test.dom</domainname>
            <allowtransfer/>
            <allowquery>02ed2632-0d50-43e5-8f81-a500a9d403c2</allowquery>
            <serial>1908271032</serial>
            <ttl>86400</ttl>
            <refresh>21600</refresh>
            <retry>3600</retry>
            <expire>3542400</expire>
            <negative>3600</negative>
            <mailadmin>mail.opnsense.localdomain</mailadmin>
            <dnsserver>opnsense.localdomain</dnsserver>
          </domain>
        </domains>
      </domain>
    </bind>
```

Also I tried to create a record for this Domain but that didn't help.
Please Help

**Comment by mimugmail** [comment]
Screenshot please :)

**Comment by blablup** [comment]
Sure.

Named.conf
```

acl "myLocalnets" { 192.168.176.0/20; };

options {

        directory       "/usr/local/etc/namedb/working";
        pid-file        "/var/run/named/pid";
        dump-file       "/var/dump/named_dump.db";
        statistics-file "/var/stats/named.stats";

         listen-on port 53530 { 127.0.0.1; 192.168.178.225; };
         listen-on-v6 port 53530 { ::1; };

        forwarders    { 192.168.178.1; };

        response-policy { zone "whitelist.localdomain"; zone "blacklist.localdomain"; };

        recursion          yes;
        allow-recursion    { myLocalnets; };

        max-cache-size    10%;
        dnssec-validation    no;
        hostname none;
        version none;
	rate-limit {
                responses-per-second 5;
                exempt-clients { 127.0.0.1; ::1; 192.168.176.0/20; };
        };
};

key "rndc-key" {
        algorithm hmac-sha256;
        secret "notforpublic";
};
controls {
        inet 127.0.0.1 port 9530
                allow { 127.0.0.1; } keys { "rndc-key"; };
};

zone "." { type hint; file "/usr/local/etc/namedb/named.root"; };

zone "localhost"        { type master; file "/usr/local/etc/namedb/master/localhost-forward.db"; };
zone "127.in-addr.arpa" { type master; file "/usr/local/etc/namedb/master/localhost-reverse.db"; };
zone "0.ip6.arpa"       { type master; file "/usr/local/etc/namedb/master/localhost-reverse.db"; };

zone "whitelist.localdomain" { type master; file "/usr/local/etc/namedb/master/whitelist.db"; notify no; check-names ignore; };
zone "blacklist.localdomain" { type master; file "/usr/local/etc/namedb/master/blacklist.db"; notify no; check-names ignore; };

zone "test.dom" { type master; file "/usr/local/etc/namedb/master/test.dom.db";  allow-query { myLocalnets; }; };


logging {
        channel default_log {
                file "/var/log/named/named.log" versions 3 size 50m;
                print-time yes;
                print-severity yes;
                print-category yes;
        };

        channel query_log {
                file "/var/log/named/query.log" versions 3 size 50m;
                print-time yes;
        };

        channel rpz_log {
                file "/var/log/named/rpz.log" versions 3 size 50m;
                print-time yes;
        };

        category default { default_log; };
        category general { default_log; };
        category queries { query_log; };
        category rpz { rpz_log; };
};
```
ls

```
root@OPNsense:~ # ls -la /usr/local/etc/namedb/master/
total 44
drwxr-xr-x  2 root  wheel  512 Aug 27 10:32 .
drwxr-xr-x  6 root  wheel  512 Aug 27 13:34 ..
-rw-r--r--  1 root  wheel  262 Aug 27 13:34 bing.db
-rw-r--r--  1 root  wheel  303 Aug 27 13:34 blacklist.db
-rw-r--r--  1 root  wheel  262 Aug 27 13:34 duckduckgo.db
-rw-r--r--  1 root  wheel  162 Aug  3 23:29 empty.db
-rw-r--r--  1 root  wheel  262 Aug 27 13:34 google.db
-rw-r--r--  1 root  wheel  172 Aug  3 23:29 localhost-forward.db
-rw-r--r--  1 root  wheel  240 Aug  3 23:29 localhost-reverse.db
-rw-r--r--  1 root  wheel    0 Aug 27 13:34 test.dom.db
-rw-r--r--  1 root  wheel  307 Aug 27 13:34 whitelist.db
-rw-r--r--  1 root  wheel  262 Aug 27 13:34 youtube.db

```

GUI:

![Screenshot from 2019-08-27 15-38-11](https://user-images.githubusercontent.com/688968/63776105-021a8500-c8e1-11e9-9514-e6231b7ba219.png)
![Screenshot from 2019-08-27 15-38-24](https://user-images.githubusercontent.com/688968/63776102-0181ee80-c8e1-11e9-8e3f-0bfea2c7d32d.png)
![Screenshot from 2019-08-27 15-39-35](https://user-images.githubusercontent.com/688968/63776100-00e95800-c8e1-11e9-8f98-8fb3201214cd.png)





I don't know the framework very well. Is there a possibility to run the template engine in debug mode?

**Comment by mimugmail** [comment]
Can't reproduce here. Can you post output from `clog -f /var/log/configd.log` while restarting BIND.
Also config.xml part after you added a record to the domain.

**Comment by mimugmail** [comment]
I can only reproduce when I disable the domain, then the file is just empty. If I have no records the file is 122kb

**Comment by blablup** [comment]
Wow you're really fast.

clog -f
```
Aug 27 14:03:59 OPNsense configd.py: [f6b4503a-2707-4f4f-a705-aef8ad3012e8] restarting BIND 
Aug 27 14:04:10 OPNsense configd.py: [7c69265e-4081-44c1-819e-24a0d27cb406] request BIND status 
Aug 27 14:04:10 OPNsense configd.py: [6088e49c-d472-4ea0-9f86-5c4ca2fa4f48] request BIND status 
```

config.xml
```
    <bind>
      <acl version="1.0.0">
        <acls>
          <acl uuid="02ed2632-0d50-43e5-8f81-a500a9d403c2">
            <enabled>1</enabled>
            <name>myLocalnets</name>
            <networks>192.168.176.0/20</networks>
          </acl>
        </acls>
      </acl>
      <dnsbl version="1.0.5">
        <enabled>1</enabled>
        <type>blp,emd,mw,pt,rw,za</type>
        <whitelists/>
        <forcesafegoogle>0</forcesafegoogle>
        <forcesafeduckduckgo>0</forcesafeduckduckgo>
        <forcesafeyoutube>0</forcesafeyoutube>
        <forcestrictbing>0</forcestrictbing>
      </dnsbl>
      <general version="1.0.5">
        <enabled>1</enabled>
        <enablerpz>1</enablerpz>
        <listenv4>127.0.0.1,192.168.178.225</listenv4>
        <listenv6>::1</listenv6>
        <port>53530</port>
        <forwarders>192.168.178.1</forwarders>
        <logsize>50</logsize>
        <maxcachesize>8</maxcachesize>
        <recursion>02ed2632-0d50-43e5-8f81-a500a9d403c2</recursion>
        <dnssecvalidation>no</dnssecvalidation>
        <hidehostname>1</hidehostname>
        <hideversion>1</hideversion>
        <enableratelimiting>1</enableratelimiting>
        <ratelimitcount>5</ratelimitcount>
        <ratelimitexcept>127.0.0.1,::1,192.168.176.0/20</ratelimitexcept>
      </general>
      <record version="1.0.0">
        <records>
          <record uuid="4c9fcb79-1206-4c89-8ee7-7550b04c67fe">
            <enabled>1</enabled>
            <domain>e38802f4-1382-4921-9c62-6a5edb98221f</domain>
            <name>test</name>
            <type>A</type>
            <value>192.168.178.225</value>
          </record>
        </records>
      </record>
      <domain version="1.0.0">
        <domains>
          <domain uuid="e38802f4-1382-4921-9c62-6a5edb98221f">
            <enabled>1</enabled>
            <type>master</type>
            <masterip/>
            <domainname>test.dom</domainname>
            <allowtransfer/>
            <allowquery>02ed2632-0d50-43e5-8f81-a500a9d403c2</allowquery>
            <serial>1908271405</serial>
            <ttl>86400</ttl>
            <refresh>21600</refresh>
            <retry>3600</retry>
            <expire>3542400</expire>
            <negative>3600</negative>
            <mailadmin>mail.opnsense.localdomain</mailadmin>
            <dnsserver>opnsense.localdomain</dnsserver>
          </domain>
        </domains>
      </domain>
    </bind>

```

**Comment by blablup** [comment]
By the way, the log of bind shows:
```

27-Aug-2019 14:20:49.536 | zoneload: error: zone test.dom/IN: not loaded due to errors.
27-Aug-2019 14:20:49.536 | zoneload: error: zone test.dom/IN: has no NS records
27-Aug-2019 14:20:49.536 | zoneload: error: zone test.dom/IN: has 0 SOA records

```

**Comment by mimugmail** [comment]
Yes, because it's empty. I imported your config and the file was empty, after enabling/disabling the domain couple of times for testing it was working ... crazy.

**Comment by blablup** [comment]
I think I found the bug. If I create a second domain, both files are filled. So this only happens if only one domain exists. 
I'm not sure what this line does maybe it is the problem?
[Line4 in template](https://github.com/opnsense/plugins/blob/8f06fc7ec03af17f0f1f6cd1655c1041bd85102f/dns/bind/src/opnsense/service/templates/OPNsense/Bind/domain.db#L4)

**Comment by mimugmail** [comment]
@AdSchellevis do we need some like within wireguard to fix this? https://github.com/opnsense/plugins/blob/8f06fc7ec03af17f0f1f6cd1655c1041bd85102f/net/wireguard/src/opnsense/service/templates/OPNsense/Wireguard/wireguard-server.conf#L5

**Comment by AdSchellevis** [comment]
@mimugmail without having looked into the template yet, if you're iterating over the data into new files, you need to filter within, yes.

**Comment by mimugmail** [comment]
Seems configd breaks when there's no record:

`Aug 27 16:44:38 OPNsense configd.py: [dd1075be-8a51-48f9-9ae1-2e4103ac3c09] Inline action failed with OPNsense/Bind OPNsense/Bind/domain.db 'collections.OrderedDict object' has no attribute 'records' at Traceback (most recent call last):   File "/usr/local/opnsense/service/modules/template.py", line 262, in _generate     content = j2_page.render(cnf_data)   File "/usr/local/lib/python3.7/site-packages/jinja2/asyncsupport.py", line 76, in render     return original_render(self, *args, **kwargs)   File "/usr/local/lib/python3.7/site-packages/jinja2/environment.py", line 1008, in render     return self.environment.handle_exception(exc_info, True)   File "/usr/local/lib/python3.7/site-packages/jinja2/environment.py", line 780, in handle_exception     reraise(exc_type, exc_value, tb)   File "/usr/local/lib/python3.7/site-packages/jinja2/_compat.py", line 37, in reraise     raise value.with_traceback(tb)   File "/usr/local/opnsense/service/modules/../templates/OPNsense/Bind/domain.db", line 8, in top-level template code     {%        
`

**Comment by mimugmail** [comment]
Forget previous .. it works now :)
@blablup fix will be in 19.7.4 (too late for .3)
